lib/charms/vault_k8s/v0/vault_client.py

```diff
@@ -0,0 +1,314 @@
+            name=role,
+            mount_point=mount,
+            extra_params={
+                "allowed_domains": allowed_domains,
+                "allow_subdomains": True,
+            },
+        )
+        logger.info("Created a role for the PKI backend")
+
+    def is_pki_role_created(self, role: str, mount: str) -> bool:
+        """Check if the role is created for the PKI backend."""
+        try:
+            existing_roles = self._client.secrets.pki.list_roles(mount_point=mount)
+            return role in existing_roles["data"]["keys"]
+        except hvac.exceptions.InvalidPath:
+            return False
+
+    def configure_kv_policy(self, policy: str, mount: str):
+        """Create/update a policy within vault to access the KV mount."""
+        with open("src/templates/kv_mount.hcl", "r") as fd:
+            mount_policy = fd.read()
+        self._client.sys.create_or_update_policy(policy, mount_policy.format(mount=mount))
+
+    def audit_device_enabled(self, device_type: str, path: str) -> bool:
+        """Check if audit device is enabled."""
+        audit_devices = self._client.sys.list_enabled_audit_devices()
+        if f"{device_type}/" not in audit_devices["data"].keys():
+            return False
+        if audit_devices["data"][f"{device_type}/"]["options"]["file_path"] != path:
+            return False
+        return True
+
+    def enable_audit_device(self, device_type: str, path: str) -> None:
+        """Enable a new audit device at the supplied path."""
+        self._client.sys.enable_audit_device(
+            device_type=device_type,
+            options={"file_path": path},
+        )
+        logger.info("Enabled audit device %s", device_type)
+
+    def create_snapshot(self) -> requests.Response:
+        """Create a snapshot of the Vault data."""
+        return self._client.sys.take_raft_snapshot()
+
+    def restore_snapshot(self, snapshot: bytes) -> requests.Response:
+        """Restore a snapshot of the Vault data.
+
+        Uses force_restore_raft_snapshot to restore the snapshot
+        even if the unseal key used at backup time is different from the current one.
+        """
+        return self._client.sys.force_restore_raft_snapshot(snapshot)
+
+    def configure_approle(self, name: str, policies: List[str]) -> str:
+        """Create/update a role within vault associating the supplied policies."""
+        self._client.auth.approle.create_or_update_approle(
+            name,
+            token_ttl="60s",
+            token_max_ttl="60s",
+            token_policies=policies,
+            bind_secret_id="true",
+        )
+        response = self._client.auth.approle.read_role_id(name)
+        return response["data"]["role_id"]
+
+    def generate_role_secret_id(self, name: str) -> str:
+        """Generate a new secret tied to an AppRole."""
+        response = self._client.auth.approle.generate_secret_id(name)
+        return response["data"]["secret_id"]
```

Comment:
`cidrs` restriction was removed from these two functions, and will need to first be merged upstream if we decide to keep it that way.